`Tabs` animation (HDS-607)

KristinLBradley opened the pull request:
### :pushpin: Summary

This PR is to add animation to the indicator bar underlining the currently selected Tab. Once approved, it will be merged into the [feature branch](https://github.com/hashicorp/design-system/pull/548).

https://hds-components-git-hds-607-tabs-animation-hashicorp.vercel.app/components/tabs#showcase

<!-- 
### :hammer_and_wrench: Detailed description
If more details are appropriate, add them here. What code changed, and why?

### :camera_flash: Screenshots
Screenshots always help, especially if this PR will change what renders to the browser -->

### :link: External links
https://hashicorp.atlassian.net/browse/HDS-607
https://www.figma.com/file/SvPCXDDW9Cv2j5isTALGNT/%F0%9F%92%80-FY23-Q3?node-id=30%3A25985

***

### 👀 How to review

<!-- Suggest how it's best for the reviewer to review the code (choose one, or remove) -->
👉 Review commit-by-commit
👉 Review by files changed

Reviewer's checklist:

- [ ] +1 Percy if applicable
- [ ] Confirm that PR has a changelog update via [Changesets](https://github.com/changesets/changesets) if needed

:speech_balloon: Please consider using [conventional comments](https://conventionalcomments.org/) when reviewing this PR.

KristinLBradley commented on packages/components/addon/components/hds/tabs/index.hbs:
```diff
@@ -0,0 +1,34 @@
+          )
+        )
+      }}
+      <li class="hds-tabs__tab-indicator" role="presentation"></li>
```
I used an li element for the tab-indicator bar. For this reason, I needed to disable the linter using: `{{! template-lint-disable no-invalid-role }}`.

The only way I could see around doing this would be to use a bunch of nested divs instead of a ul list to mark up the tabs. (The individual tabs themselves need to be contained in a parent element with the "tablist" role. The styling which changes the tab-indicator color on hover relies on it being a sibling of the tabs otherwise I could probably use a pseudo selector for the tab-indicator instead.)

Let me know if you don't like this or what you think.

alex-ju replied:
The only issue I would've had with it was if the number of items announced by screen readers was altered because of it, but as we use `role="presentation"` the indicator should not be part of the accessibility tree, so no problem there.

KristinLBradley replied:
Thanks, merging this branch for now. I can continue to iterate on this if there are other proposals though.

KristinLBradley commented on packages/components/tests/integration/components/hds/tabs/index-test.js:
```diff
@@ -7,7 +7,14 @@ module('Integration | Component | hds/tabs/index', function (hooks) {
+        <T.Panel data-test="panel">Content 1</T.Panel>
+        <T.Panel>Content 2</T.Panel>
+      </Hds::Tabs>
+  `);
```
The test was failing before I added the child elements

didoo replied:
OK

(in this case, you can remove the `data-test` attributes, are not used)

KristinLBradley commented on packages/components/tests/dummy/app/templates/components/tabs.hbs:
```diff
@@ -284,10 +280,10 @@
         </ul>
         <ul class="dummy-tab-ul-tab-wrapper">
-          <Hds::Tabs::Tab @icon="hexagon" @count="10" mock-state-value={{state}} mock-state-selector="a">Lorem ipsum</Hds::Tabs::Tab>
+          <Hds::Tabs::Tab @icon="hexagon" @count="10" mock-state-value={{state}} mock-state-selector="*">Lorem ipsum</Hds::Tabs::Tab>
```
The Tab DOM was refactored to use a button element instead of a link. The Icon and Count had to be moved outside of the button element to fix accessibility issues the linter complained about. In order to change the icon color on hover as well as the button color, they both needed to use a mock-hover class. (I couldn't figure out how to add it the parent wrapper component instead, using "li" as the selector didn't work. If there's a better way to do this please let me know.)

didoo replied:
@KristinLBradley I have pushed a commit (https://github.com/hashicorp/design-system/pull/607/commits/b34dc98a11959cc21663daf17b4191d87590ea9c) that fixes this. Have a look at my changes and let me know if something is not clear (in this case, it's slightly different from other components, because the elements that receive the `hover` and the `focus` states are different).

KristinLBradley commented on packages/components/addon/components/hds/tabs/index.js:
```diff
@@ -79,6 +81,21 @@ export default class HdsTabsIndexComponent extends Component {
+      tabLeftPos + 'px'
+    );
+    tabsParentElem.style.setProperty('--indicatior-width', tabWidth + 'px');
+  }
```
Let me know if this violates any Ember best practices and/or if there's better ways to do this.

KristinLBradley commented:
@heatherlarsen Please review the animation behavior and overall visual styles: https://hds-components-git-hds-607-tabs-animation-hashicorp.vercel.app/components/tabs#showcase Thanks!

didoo commented on packages/components/app/styles/components/tabs.scss:
```diff
@@ -86,3 +81,25 @@ $indicator-height: 3px;
+  position: absolute;
+  right: 0;
+  // adjust position to make indicator overlap tabs bottom border:
+  bottom: floor(calc($indicator-height / 2));
```
[issue] checking how it's designed in Figma, I think this value should be 3px (`$indicator-height`)
```suggestion
  bottom: $indicator-height;
```
<img width="1349" alt="screenshot_1852" src="https://user-images.githubusercontent.com/686239/194056438-49025857-93f2-4fc8-b9d0-8015c12a6f5f.png">

KristinLBradley replied:
I had deleted styles positioning the tabs when I was refactoring. I added them back so that the focus outline should now sit right above the indicator underline again. (One complicating factor is that the indicator needs to overlap the bottom gray border.)

didoo reviewed (commented):
added a few minor comments. 

once fixed the typo, I think we're good

KristinLBradley commented on packages/components/addon/components/hds/tabs/index.js:
```diff
@@ -79,6 +81,21 @@ export default class HdsTabsIndexComponent extends Component {
+
+    // Set CSS custom properties for indicator
+    tabsParentElem.style.setProperty(
+      '--indicatior-left-pos',
```
At least I made it consistently ;)

heatherlarsen commented:
It sounds like the animation was a little complex with the positioning of the indicator (apologies for that), but it looks really wonderful 🎉 I tested reduced-motion and that worked great as well!

A couple of things I noticed about the docs, in general: 
1. In the showcase each individual tab has a light gray border around it. I wonder if folks will assume that's a part of the component itself and get confused when it doesn't appear in the fully rendered version. In other component showcases, each atomic element stands on its own and isn't contained by a border, wdyt about removing them to keep it consistent?
2. The design guidelines screenshot appears to just be of the stickersheet. This is usually of the [Design Guidelines from Figma](https://www.figma.com/file/noyY6dUMDYjmySpHcMjhkN/HDS-Product---Components?node-id=17023%3A45118). I can grab you the screenshot if you'd like, but it might be best to have @didoo show you his process for doing this so there's not a dependency on Design in the future. Let me know which you'd prefer. 

Let me know if these are out of scope of this PR; I can move them into issues.

didoo commented on packages/components/app/styles/components/tabs.scss:
```diff
@@ -86,3 +85,23 @@ $indicator-height: 3px;
+.hds-tabs__tab-indicator {
+  position: absolute;
+  right: 0;
+  bottom: $indicator-height;
+  bottom: 0;
```
notice: there are two bottom declarations
```suggestion
  bottom: 0;
```

KristinLBradley replied:
Sorry I missed that when I was refactoring. I will clean up

didoo reviewed (approved):
just a note about a duplicated property. apart from that, all good for me

KristinLBradley commented:
> > * The light gray outline around individual Tab components in the showcase was added by @didoo Should I delete those?
> 
> @KristinLBradley @heatherlarsen the gray border is used to highlight the layout of the `Tab` element. But now I remembered that we had something similar for the Form base controls and we ended up using a toggle (see "Field" in the showcase: https://design-system-components-hashicorp.vercel.app/components/form/base-elements#showcase). So Kristin, you can remove the border and I'll open a new PR to add the toggle (unless you want to try to do it yourself, in the same way is done in the example linked above)
> 
> > * If @didoo has a recommended process for grabbing the Figma screenshot I'd be happy to follow that & replace the current image.
> 
> I can show you tomorrow how to do it (we can probably to it in a follow-up PR).

Thanks. For now I deleted the outline style adding the gray border around Tabs in docs and added TODO comment in the doc file for replacing the screenshot and adding an outline toggle.
